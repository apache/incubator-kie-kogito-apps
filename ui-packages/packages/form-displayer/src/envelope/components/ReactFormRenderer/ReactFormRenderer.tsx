/*
 * Copyright 2021 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useEffect, useState } from 'react';

import uuidv4 from 'uuid';
import * as Babel from '@babel/standalone';
import ReactDOM from 'react-dom';
import * as Patternfly from '@patternfly/react-core';
import { FormArgs } from '../../../api';

interface ReactFormRendererProps {
  content: FormArgs;
}

const ReactFormRenderer: React.FC<ReactFormRendererProps> = ({ content }) => {
  const [source, setSource] = useState<string>();

  useEffect(() => {
    if (content && content.source) {
      setSource(content.source['source-content']);
      renderform();
    }
  }, [content]);

  const renderform = () => {
    if (source) {
      window.React = React;
      window.ReactDOM = ReactDOM;

      // @ts-ignore
      window.PatternFlyReact = Patternfly;

      const container: HTMLElement = document.getElementById('formContainer');
      container.innerHTML = '';
      const id = uuidv4();
      const formContainer: HTMLElement = document.createElement('div');
      formContainer.id = id;

      container.appendChild(formContainer);

      const reactReg = /import React, {[^}]*}.*(?='react').*/gim;
      const patternflyReg = /import {[^}]*}.*(?='@patternfly\/react-core').*/gim;
      const regexvalueReact = new RegExp(reactReg);
      const reactImport = regexvalueReact.exec(source);
      const reg = /\{([^)]+)\}/;
      const reactElements = reg.exec(reactImport[0])[1];
      console.log('react', reactElements);
      const regexvaluePat = new RegExp(patternflyReg);
      const patternflyImport = regexvaluePat.exec(source);
      const patternflyElements = reg.exec(patternflyImport[0])[1];
      console.log('pat', patternflyElements);
      const trimmedSource = source
        .split(reactReg)
        .join('')
        .trim()
        .split(patternflyReg)
        .join('')
        .trim();
      const tempSource = trimmedSource;
      const formName = tempSource.split(':')[0].split('const ')[1];
      try {
        const react = Babel.transform(trimmedSource, {
          presets: [
            'react',
            [
              'typescript',
              {
                allExtensions: true,
                isTSX: true
              }
            ]
          ]
        }).code;

        const compiledReact = react;

        const scriptElement: HTMLScriptElement = document.createElement(
          'script'
        );

        // @ts-ignore
        window.PatternFly = window.PatternFlyReact;

        scriptElement.type = 'module';

        const content = `
        const {${reactElements}} = React;
        const {${patternflyElements}} = PatternFlyReact;
       
        
        ${compiledReact}
        const target = document.getElementById("${id}");
        const element = window.React.createElement(${formName}, {});
        window.ReactDOM.render(element, target);
        `;
        console.log('cone', content);
        scriptElement.text = content;

        container.appendChild(scriptElement);
      } catch (e) {
        console.error(e);
      }
    }
  };

  return (
    <div
      style={{
        height: '100%'
      }}
      id={'formContainer'}
    >
      {}
    </div>
  );
};

export default ReactFormRenderer;
