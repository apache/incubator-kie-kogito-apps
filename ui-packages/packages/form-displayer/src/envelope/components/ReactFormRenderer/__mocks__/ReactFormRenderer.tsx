import React from 'react';

const ReactFormRenderer = () => {
  return <></>;
};

export default ReactFormRenderer;
