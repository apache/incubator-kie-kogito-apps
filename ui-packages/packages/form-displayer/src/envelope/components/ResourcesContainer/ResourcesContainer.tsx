import React, { ReactElement } from 'react';
import { Helmet } from 'react-helmet';
import { FormResources } from '../../../api';

interface ResourcesContainerProps {
  resources: FormResources;
}

const ResourcesContainer: React.FC<ResourcesContainerProps> = ({
  resources
}) => {
  const scriptTags: ReactElement[] = [];
  const styleTags: ReactElement[] = [];

  if (resources) {
    for (const key in resources.styles) {
      styleTags.push(
        <link rel={'stylesheet'} href={resources.styles[key]} key={key} />
      );
    }
    for (const key in resources.scripts) {
      scriptTags.push(
        <script
          src={resources.scripts[key]}
          type={'text/javascript'}
          async={false}
          key={key}
        />
      );
    }
  }

  return (
    <Helmet>
      {styleTags}
      {scriptTags}
    </Helmet>
  );
};

export default ResourcesContainer;
