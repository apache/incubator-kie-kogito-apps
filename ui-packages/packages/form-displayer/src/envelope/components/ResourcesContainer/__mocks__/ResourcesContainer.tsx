import React from 'react';

const ResourcesContainer = () => {
  return <></>;
};

export default ResourcesContainer;
