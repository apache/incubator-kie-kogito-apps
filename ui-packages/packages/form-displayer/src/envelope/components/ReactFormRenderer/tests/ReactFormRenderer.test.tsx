/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import React from 'react';
import { mount } from 'enzyme';
import ReactFormRenderer from '../ReactFormRenderer';
import ResourcesContainer from '../../ResourcesContainer/ResourcesContainer';

jest.mock('../../ResourcesContainer/ResourcesContainer');

describe('ReactFormRenderer component tests', () => {
  beforeAll(() => {
    const div = document.createElement('div');
    div.setAttribute('id', 'formContainer');
    document.body.appendChild(div);
  });

  it('Snapshot test with default props', () => {
    const props = {
      source:
        "import React, { useState } from 'react';\nimport {\n  Card,\n  CardBody,\n  TextInput,\n  FormGroup,\n  Checkbox\n} from '@patternfly/react-core';\n\nconst Form__hiring_HRInterview: React.FC<any> = (props: any) => {\n  const [candidate__email, set__candidate__email] = useState<string>();\n  const [candidate__name, set__candidate__name] = useState<string>();\n  const [candidate__salary, set__candidate__salary] = useState<string>();\n  const [candidate__skills, set__candidate__skills] = useState<string>();\n  const [approve, set__approve] = useState<boolean>();\n\n  return (\n    <div className={'pf-c-form'}>\n      <Card>\n        <CardBody className=\"pf-c-form\">\n          <label>\n            <b>Candidate</b>\n          </label>\n          <FormGroup\n            fieldId={'uniforms-0000-0002'}\n            label={'Email'}\n            isRequired={false}\n          >\n            <TextInput\n              name={'candidate.email'}\n              id={'uniforms-0000-0002'}\n              isDisabled={true}\n              placeholder={''}\n              type={'text'}\n              value={candidate__email}\n              onChange={set__candidate__email}\n            />\n          </FormGroup>\n          <FormGroup\n            fieldId={'uniforms-0000-0003'}\n            label={'Name'}\n            isRequired={false}\n          >\n            <TextInput\n              name={'candidate.name'}\n              id={'uniforms-0000-0003'}\n              isDisabled={true}\n              placeholder={''}\n              type={'text'}\n              value={candidate__name}\n              onChange={set__candidate__name}\n            />\n          </FormGroup>\n          <FormGroup\n            fieldId={'uniforms-0000-0005'}\n            label={'Salary'}\n            isRequired={false}\n          >\n            <TextInput\n              type={'number'}\n              name={'candidate.salary'}\n              isDisabled={true}\n              id={'uniforms-0000-0005'}\n              placeholder={''}\n              step={1}\n              value={candidate__salary}\n              onChange={set__candidate__salary}\n            />\n          </FormGroup>\n          <FormGroup\n            fieldId={'uniforms-0000-0006'}\n            label={'Skills'}\n            isRequired={false}\n          >\n            <TextInput\n              name={'candidate.skills'}\n              id={'uniforms-0000-0006'}\n              isDisabled={true}\n              placeholder={''}\n              type={'text'}\n              value={candidate__skills}\n              onChange={set__candidate__skills}\n            />\n          </FormGroup>\n        </CardBody>\n      </Card>\n      <FormGroup fieldId=\"uniforms-0000-0008\">\n        <Checkbox\n          isChecked={approve}\n          isDisabled={false}\n          id={'uniforms-0000-0008'}\n          name={'approve'}\n          label={'Approve'}\n          onChange={set__approve}\n        />\n      </FormGroup>\n    </div>\n  );\n};\n\nexport default Form__hiring_HRInterview;\n",
      resources: {
        scripts: {
          'bootstrap.min.js':
            'https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/js/bootstrap.min.js',
          'jquery.js': 'https://code.jquery.com/jquery-3.2.1.slim.min.js',
          'popper.js':
            'https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.12.9/umd/popper.min.js'
        },
        styles: {
          'bootstrap.min.css':
            'https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css'
        }
      },
      setIsExecuting: jest.fn()
    };
    const wrapper = mount(<ReactFormRenderer {...props} />, {
      attachTo: document.getElementById('formContainer')
    });

    wrapper.update();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('div')).toBeTruthy();

    const resources = wrapper.find(ResourcesContainer);
    expect(resources.exists()).toBeTruthy();
  });
});
