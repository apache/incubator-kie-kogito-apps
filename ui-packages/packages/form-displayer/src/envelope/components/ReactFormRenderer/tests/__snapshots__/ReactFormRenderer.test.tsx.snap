// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactFormRenderer component tests Snapshot test with default props 1`] = `
<ReactFormRenderer
  resources={
    Object {
      "scripts": Object {
        "bootstrap.min.js": "https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/js/bootstrap.min.js",
        "jquery.js": "https://code.jquery.com/jquery-3.2.1.slim.min.js",
        "popper.js": "https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.12.9/umd/popper.min.js",
      },
      "styles": Object {
        "bootstrap.min.css": "https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css",
      },
    }
  }
  setIsExecuting={
    [MockFunction] {
      "calls": Array [
        Array [
          true,
        ],
        Array [
          false,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  source="import React, { useState } from 'react';
import {
  Card,
  CardBody,
  TextInput,
  FormGroup,
  Checkbox
} from '@patternfly/react-core';

const Form__hiring_HRInterview: React.FC<any> = (props: any) => {
  const [candidate__email, set__candidate__email] = useState<string>();
  const [candidate__name, set__candidate__name] = useState<string>();
  const [candidate__salary, set__candidate__salary] = useState<string>();
  const [candidate__skills, set__candidate__skills] = useState<string>();
  const [approve, set__approve] = useState<boolean>();

  return (
    <div className={'pf-c-form'}>
      <Card>
        <CardBody className=\\"pf-c-form\\">
          <label>
            <b>Candidate</b>
          </label>
          <FormGroup
            fieldId={'uniforms-0000-0002'}
            label={'Email'}
            isRequired={false}
          >
            <TextInput
              name={'candidate.email'}
              id={'uniforms-0000-0002'}
              isDisabled={true}
              placeholder={''}
              type={'text'}
              value={candidate__email}
              onChange={set__candidate__email}
            />
          </FormGroup>
          <FormGroup
            fieldId={'uniforms-0000-0003'}
            label={'Name'}
            isRequired={false}
          >
            <TextInput
              name={'candidate.name'}
              id={'uniforms-0000-0003'}
              isDisabled={true}
              placeholder={''}
              type={'text'}
              value={candidate__name}
              onChange={set__candidate__name}
            />
          </FormGroup>
          <FormGroup
            fieldId={'uniforms-0000-0005'}
            label={'Salary'}
            isRequired={false}
          >
            <TextInput
              type={'number'}
              name={'candidate.salary'}
              isDisabled={true}
              id={'uniforms-0000-0005'}
              placeholder={''}
              step={1}
              value={candidate__salary}
              onChange={set__candidate__salary}
            />
          </FormGroup>
          <FormGroup
            fieldId={'uniforms-0000-0006'}
            label={'Skills'}
            isRequired={false}
          >
            <TextInput
              name={'candidate.skills'}
              id={'uniforms-0000-0006'}
              isDisabled={true}
              placeholder={''}
              type={'text'}
              value={candidate__skills}
              onChange={set__candidate__skills}
            />
          </FormGroup>
        </CardBody>
      </Card>
      <FormGroup fieldId=\\"uniforms-0000-0008\\">
        <Checkbox
          isChecked={approve}
          isDisabled={false}
          id={'uniforms-0000-0008'}
          name={'approve'}
          label={'Approve'}
          onChange={set__approve}
        />
      </FormGroup>
    </div>
  );
};

export default Form__hiring_HRInterview;
"
>
  <div>
    <ResourcesContainer
      resources={
        Object {
          "scripts": Object {
            "bootstrap.min.js": "https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/js/bootstrap.min.js",
            "jquery.js": "https://code.jquery.com/jquery-3.2.1.slim.min.js",
            "popper.js": "https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.12.9/umd/popper.min.js",
          },
          "styles": Object {
            "bootstrap.min.css": "https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css",
          },
        }
      }
    />
    <div
      id="formContainer"
      style={
        Object {
          "height": "100%",
        }
      }
    />
  </div>
</ReactFormRenderer>
`;
